ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java

```diff
@@ -485,6 +480,26 @@ private void clearLocksAndHB() {
     stopHeartbeat();
   }
 
+  private void cleanupDirForCTAS() {
+    if (HiveConf.getBoolVar(conf, HiveConf.ConfVars.TXN_CTAS_X_LOCK)) {
```

Comment:
shouldn't we check that this was a CTAS operation? do we only need to cleanup if an exclusive lock is enabled?

Reply:
The idea is that `destinationTable` object is set only when CTAS operations are performed ([here](https://github.com/apache/hive/pull/3457/files#diff-d4b1a32bbbd9e283893a6b52854c7aeb3e356a1ba1add2c4107e52901ca268f9R7856)). So no inherent need of checking whether the operation is CTAS.

As far as exclusive lock is enabled, it would be right to perform cleanup when a exclusive lock is acquired otherwise we might have a situation wherein the cleaner is cleaning and concurrent CTAS operations write to the same location causing issues.

Reply:
shouldn't we do cleanup when there is no concurrent CTAS, but the operation was canceled by the user in the middle?

Reply:
Yes, the cleanup is for cases when there is no concurrent CTAS and a single query fails. But if we disable exclusive locking and perform concurrent CTAS operations and let's say the first query fails, then cleanup is triggered by the first query on the same location as well the second query will write to the same location. 

This is the situation I want to avoid which is why perform cleanup only when the exclusive locking on CTAS is enabled.